vault/identity_store_injector_testonly.go

```diff
@@ -501,58 +500,64 @@ func (i *IdentityStore) CreateDuplicateGroupsInStorage(ctx context.Context, flag
-// CreateDuplicateEntityAliasesInStorage creates n entities with a duplicate alias in storage
-// This should only be used in testing
+// CreateDuplicateEntityAliasesInStorage creates n entities with a duplicate
+// alias in storage This should only be used in testing. This method can only
+// create non-local aliases. Local aliases are stored differently.
 //
 // Pass in mount type and accessor to create the entities
-func (i *IdentityStore) CreateDuplicateEntityAliasesInStorage(ctx context.Context, flags DuplicateEntityAliasFlags) ([]string, []string, error) {
-	var bucketKeys []string
+func (i *IdentityStore) CreateDuplicateEntityAliasesInStorage(ctx context.Context, flags DuplicateEntityAliasFlags) ([]string, error) {
 	var entityIDs []string
-
+	if flags.NamespaceID == "" {
+		flags.NamespaceID = namespace.RootNamespaceID
+	}
 	for d := 0; d < flags.Count; d++ {
-		entityID := fmt.Sprintf("%s-%d", flags.Name, d)
-
-		policyID := fmt.Sprintf("policy-%s-%d", flags.Name, d)
+		aliasID, err := uuid.GenerateUUID()
+		if err != nil {
+			return nil, err
+		}
 
-		entityDupName := fmt.Sprintf("%s-entity-dup-%d", flags.Name, d)
-		aliasDupName := fmt.Sprintf("%s-alias-dup", flags.Name)
+		// Alias name is either exact match or different case
+		dupAliasName := flags.Name
+		if flags.DifferentCase {
+			dupAliasName = randomCase(flags.Name)
+		}
+
+		// In real life alias dupes are due to races where they were auto-created
+		// along with the entities they point to. When we auto create entities like
+		// this they get random names so never collide with each other directly so
+		// don't create entities with duplicate names for this case as it doesn't
+		// match what customers who get in this state see. Instead use the same code
+		// path as CreateOfFetchEntity.
+		e := new(identity.Entity)
+		e.NamespaceID = flags.NamespaceID
+		err = i.sanitizeEntity(ctx, e)
+		if err != nil {
+			return nil, err
+		}
+		entityIDs = append(entityIDs, e.ID)
 
 		a := &identity.Alias{
-			ID:            entityID,
-			CanonicalID:   entityID,
+			ID:            aliasID,
+			CanonicalID:   e.ID,
 			MountAccessor: flags.CommonAliasFlags.MountAccessor,
-			Name:          aliasDupName,
-		}
-
-		bucketKey := i.entityPacker.BucketKey(entityID)
-		bucketKeys = append(bucketKeys, bucketKey)
-		entityIDs = append(entityIDs, entityID)
-
-		e := &identity.Entity{
-			ID:   entityID,
-			Name: entityDupName,
-			Aliases: []*identity.Alias{
-				a,
-			},
-			NamespaceID: namespace.RootNamespaceID,
-			BucketKey:   bucketKey,
-			Policies:    []string{policyID},
+			Name:          dupAliasName,
 		}
+		e.UpsertAlias(a)
 
 		entity, err := ptypes.MarshalAny(e)
 		if err != nil {
-			return nil, nil, err
+			return nil, err
 		}
 		item := &storagepacker.Item{
 			ID:      e.ID,
 			Message: entity,
 		}
 		if err = i.entityPacker.PutItem(ctx, item); err != nil {
-			return nil, nil, err
+			return nil, err
 		}
 	}
 
-	return entityIDs, bucketKeys, nil
+	return entityIDs, nil
```

Comment:
This change is restoring the newer change from https://github.com/hashicorp/vault-enterprise/pull/7266 but this function was left out when those changes were adopted and merged in #29461. I _think_ the rationale for not including these at the time was because some tests used the bucket keys still. None do now and this pattern is not only more realistic but necessary for reproducing certain issues.